enhancements/oc/olm-mirroring.md

```diff
@@ -0,0 +1,115 @@
+```
+
+`oc catalog build` will:
+
+- Use appregistry protocol to retrieve all of the operators in a namespace of Quay.io specified by `appregistry-namespace`
+- Load all of the downloaded operators into a versioned `operator-registry` sqlite artifact
+- Build a runnable operator-registry image by appending the database to the `operator-registry` base image using the same machinary as `oc image append`
+- Mirror that image to the target defined by `--to`, using the same machinary as `oc image append`
+
+It will have the following flags:
+
+- `--from=ref` - the base image to add the built operator database into. Defaults to the operator-registry image shipped with the version of OpenShift that `oc` came from.
+- `--to=ref` - the location that the image will be mirrored.
+- `--save=path` - instead of mirroring, this will output a tarball containing the image
+- `--auth-token=string` - the auth token ([instructions](https://github.com/operator-framework/operator-courier#authentication)) for authenticating with appregistry.
+- `--appregistry-endpoint=url` - the CNR endpoint to authenticate against. Defaults to `"https://quay.io/cnr"`, the endpoint used by OpenShift 4.1-4.3.
+- `--appregistry-namespace=string` - the namespace in appregistry to mirror. Each repository within the namespace represents one operator. 
+- `--db-only` - if true, and `--save` is specified, the operator database will be saved in the directory noted by `--save`.
+
+This command generates and mirrors in one step, because we do not assume that there is any registry available aside from the target disconnected registry. `--save` can be used to save the image locally without a registry available.
+
+### Extract the contents of a catalog for mirroring
+
+`oc` will be extended with a second command:
+
+```sh
+$ oc catalog mirror --from=quay.io/ecordell/community-operators-catalog:4.2.1 --to=localhost:5000/community-operators --manifests=./mirror-manifests
+mirroring ...
+
+$ ls ./mirror-manifests
+imagecontentsourcepolicy.yaml
+catalogsource.yaml
+```
+
+`oc catalog mirror` will:
+
+- Pull the catalog image referenced by `--from`, using `oc` machinery
+- Read the database to get the list of operator and operand images
+- Build a mapping of images to the disconnected registry
+- Mirror all of the referenced images by (interally) calling `oc image mirror`
+- Output a set of manifests that, if applied to a cluster that has access to the mirrored images, will correctly configure nodes and OLM to use those images.
+
+It has the following flags:
+
+- `--from=ref` - a catalog image (such as one built by `oc catalog build`)
+- `--from-archive=path` - a reference to a tarball image, such as one built with `oc catalog build --save=path`
+- `--to=ref` - optional; the location that the image will be mirrored. This should include only a registry and a namespace that images can be mirrored into.
```

Comment:
what is the output if `--to` is not included?

Reply:
I think it makes sense to make this field required (this is a holdover from an iteration that just printed out the mapping).